index/segment/mem/segment.go

```diff
@@ -22,224 +22,246 @@ package mem
-			// Index already contains this document.
-			return nil
-		}
-	} else {
-		uuid, err := s.newUUIDFn()
-		if err != nil {
-			return err
-		}
-		d.Fields = append(d.Fields, doc.Field{
-			Name:  doc.IDReservedFieldName,
-			Value: uuid,
-		})
-	}
-
-	// TODO: Consider supporting concurrent writes by relaxing the requirement that
-	// inserted documents are immediately searchable.
-	s.ids.Lock()
+	{
+		s.writer.Lock()
 
-	newID := s.ids.writer.Inc()
-	s.insertDoc(newID, d)
-	err = s.insertTerms(newID, d)
-	s.ids.reader.Inc()
+		s.insertDocWithLocks(d)
+		s.readerID.Inc()
 
-	s.ids.Unlock()
+		s.writer.Unlock()
+	}
 
-	return err
+	return d.ID, nil
 }
 
-func (s *segment) Reader() (index.Reader, error) {
+func (s *segment) Batch(ds []doc.Document) error {
 	s.state.RLock()
+	defer s.state.RUnlock()
 	if s.state.closed {
-		s.state.RUnlock()
-		return nil, sgmt.ErrClosed
+		return sgmt.ErrClosed
 	}
 
-	maxID := s.ids.reader.Load()
-	r := newReader(s, maxID)
-	return r, nil
-}
+	if err := s.prepareDocs(ds); err != nil {
+		return err
+	}
 
-func (s *segment) Seal() error {
-	s.state.Lock()
-	if s.state.sealed {
-		s.state.Unlock()
-		return errSegmentSealed
+	{
+		s.writer.Lock()
+
+		for _, d := range ds {
+			s.insertDocWithLocks(d)
+		}
+		s.readerID.Add(uint32(len(ds)))
+
+		s.writer.Unlock()
 	}
 
-	s.state.sealed = true
-	s.state.Unlock()
 	return nil
 }
 
-func (s *segment) Close() error {
-	s.state.Lock()
-	if s.state.closed {
-		s.state.Unlock()
-		return sgmt.ErrClosed
+func (s *segment) prepareDocs(ds []doc.Document) error {
+	ids := idsgen.New(len(ds))
+	for i := 0; i < len(ds); {
+		d := ds[i]
+		err := d.Validate()
+		if err != nil {
+			return err
+		}
+
+		if d.HasID() {
+			if s.termsDict.ContainsTerm(doc.IDReservedFieldName, d.ID) {
+				// The segment already contains this document so we can remove it from those
+				// we need to index.
+				ds[i], ds[len(ds)] = ds[len(ds)], ds[i]
+				ds = ds[:len(ds)-1]
+				continue
+			}
+
+			if _, ok := ids.Get(d.ID); ok {
+				return errDuplicateID
+			}
+		} else {
+			id, err := s.newUUIDFn()
+			if err != nil {
+				return err
+			}
+			d.ID = id
+
+			// Update the document since we added an ID.
+			ds[i] = d
+		}
+
+		ids.SetUnsafe(d.ID, struct{}{}, idsgen.SetUnsafeOptions{
+			NoCopyKey:     true,
+			NoFinalizeKey: true,
+		})
+		i++
 	}
 
-	s.state.sealed = true
-	s.state.closed = true
-	s.state.Unlock()
 	return nil
 }
 
-func (s *segment) insertDoc(id postings.ID, d doc.Document) {
+// insertDocWithLocks inserts a document into the index. It must be called with the
+// state and writer locks.
+func (s *segment) insertDocWithLocks(d doc.Document) {
+	nextID := s.writer.nextID
+	s.indexDocWithLock(nextID, d)
+	s.storeDocWithLock(nextID, d)
+	s.writer.nextID++
+}
+
+// indexDocWithLock indexes the fields of a document in the segment's terms dictionary. It
+// must be called with the segment's state lock.
+func (s *segment) indexDocWithLock(id postings.ID, d doc.Document) error {
```

Comment:
how bout renaming to `indexDocWithStateRLock` to better indicate what lock you need. 

PS this is why I really want to create https://github.com/m3db/build-tools/issues/14